### ISSUE TITLE
Receiver of a non-pure method call can be null after the call

### ISSUE THREAD

**Issue by AndrewShf** [eisop/checker-framework#300]
A more self-contained example is pasted in the comment area and provided by wmdietl.

A minimal example is pasted below:
```
import org.checkerframework.dataflow.cfg.node.Node;
import org.checkerframework.checker.nullness.qual.Nullable;
import org.checkerframework.dataflow.cfg.node.AssignmentNode;

class PossibleNullConjunction {
    @Nullable Node currentNode;

    void test(Node n) {
        if (currentNode == null) {
            return;
        }
        if (!currentNode.getOperands().contains(n)
                && !currentNode.getTransitiveOperands().contains(n)) {
            if (n instanceof AssignmentNode) {
                Node expr = ((AssignmentNode) n).getExpression();
                Node target = ((AssignmentNode) n).getTarget();
                if (currentNode.getOperands().contains(n)
                        || currentNode.getTransitiveOperands().contains(n)) {
                    return;
                }
            }
            return;
        }
    }
}
```
Commands:
`javacheck -cp "dataflow/build/classes/java/main:/Users/Andrew/Desktop/Lab/projects/plume-util/src/main/java" -processor NullnessChecker PossibleNullConjunction.java`

Expected: Report an error on `!currentNode.getTransitiveOperands().contains(n)` after the conjunction symbol, as `getOperands()` and `getTransitiveOperands()` are not marked as `pure` methods.

Output: Get an error on `if (currentNode.getOperands().contains(n)` after the first if statement. This gives some inconsistency, why not report the error in the first place?

**Comment by AndrewShf** [comment]
I tried to delete statements related to `AssignmentNode`, but without these statements, no error will be reported. I don't know why.

For example, this piece of code only generates an implicit compilation warning.
```
import org.checkerframework.dataflow.cfg.node.Node;
import org.checkerframework.checker.interning.qual.InternedDistinct;
import org.checkerframework.checker.nullness.qual.Nullable;
import org.checkerframework.dataflow.cfg.node.AssignmentNode;

class PossibleNullConjunction {
    @InternedDistinct @Nullable Node currentNode;

    void test(Node n) {
        if (currentNode == null) {
            return;
        }
        if (!currentNode.getOperands().contains(n)
                && !currentNode.getTransitiveOperands().contains(n)) {
            if (n instanceof AssignmentNode) {
                if (currentNode.getOperands().contains(n)
                        || currentNode.getTransitiveOperands().contains(n)) {
                    return;
                }
            }
            return;
        }
    }
}
```
Output: `warning: Implicitly compiled files were not subject to annotation processing.`
Expected: Still an error on line `&& !currentNode.getTransitiveOperands().contains(n)`

**Comment by wmdietl** [comment]
I think the following is a more self-contained example of the same issue:

````java
import org.checkerframework.checker.nullness.qual.Nullable;

public class Issue300 {
    class Bug {
        void setFieldNull() {
            Issue300.this.currentNode = null;
        }
    }

    @Nullable Bug currentNode = new Bug();

    void test() {
        if (currentNode == null) {
            return;
        }
        currentNode.setFieldNull();
        currentNode.toString();
    }

    public static void main(String[] args) {
        new Issue300().test();
    }
}
````

Using the inner class was just the easiest to write - one could use some other way to get a reference back to the object that contains the field.

I think the problem is that after a de-reference, we assume that the receiver is non-null:
https://github.com/eisop/checker-framework/blob/5c0e6f9398259d76f037483f2d4ff2d5ce5d38fb/checker/src/main/java/org/checkerframework/checker/nullness/NullnessTransfer.java#L361

We should carefully think about all the `makeNonNull` calls, e.g. a field access could trigger a class initializer, which could set the receiver if the field access to null. Could you try writing such an example?

**Comment by AndrewShf** [comment]
> We should carefully think about all the `makeNonNull` calls, e.g. a field access could trigger a class initializer, which could set the receiver if the field access to null. Could you try writing such an example?

You already provide one example, do you mean writing another example without using inner class?

**Comment by AndrewShf** [comment]
Got it.

**Comment by AndrewShf** [comment]
I am still a bit confused. Did we focus more on, why the error is not reported after the conjunction and reported in the second if statement in today's discussion?

If after de-referencing, we assume the receiver is non-null, what causes the `dereference of possible nullable` on the second if statement?

**Comment by wmdietl** [comment]
In the original example, there is a non-pure call on `n`, which correctly loses the information about `currentNode`.
You couldn't reproduce the problem without the non-pure calls on `n`.
If there are only calls on `currentNode`, dataflow incorrectly continues to believe it is non-null, even though the non-pure method could reset the field.
Does this make more sense now?

**Comment by AndrewShf** [comment]
>  Does this make more sense now?
Thanks, I understand it now. `n.getExpression` could change the nullness property of `currentNode`.

**Comment by AndrewShf** [comment]
During today's discussion, we found two new false-negatives.

```java
import org.checkerframework.checker.nullness.qual.Nullable;

class Issue300B {
    @Nullable Object f = "";
    void m(Object o) {
        f = null;
    }

    public static void main(String[] args) {
        Issue300B r = new Issue300B();
        if (r.f == null) {return;}

        r.m(r.f);
        r.f.toString(); // NPE when running
    }

}

```

```java
import org.checkerframework.checker.nullness.qual.Nullable;
import org.checkerframework.dataflow.qual.*;

public final class Issue300C {
    @Nullable Issue300C f;

    Issue300C() {
        this.f = this;
    }

    void m2() {
        f = null;
    }

    @Pure
    @Nullable
    Issue300C getF() {
        return f;
    }

    public static void main(String[] args) {
        Issue300C r = new Issue300C();

        if (r.getF() != null) {
            r.getF().m2();
            r.getF().toString(); // NPE when running
        }
    }
}

```

Both issue an NPE when running, but checker-framework doesn't issue any warning.
